conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureUrlDecodingHandler.java

```diff
@@ -0,0 +1,108 @@
+    }
+
+    @Override
+    public void handleRequest(final HttpServerExchange exchange) throws Exception {
+        if (shouldDecode(exchange)) {
+            final StringBuilder sb = getStringBuilderForDecoding(exchange);
+            decodePath(exchange, charset, sb);
+            decodeQueryString(exchange, charset, sb);
+            decodePathTemplateMatch(exchange, charset, sb);
+        }
+        next.handleRequest(exchange);
+    }
+
+    // Returns true if the exchange should be decoded.  This method updates the ALREADY_DECODED
+    // attachment so that subsequent invocations will always return false.
+    private static boolean shouldDecode(final HttpServerExchange exchange) {
+        return !exchange.getConnection().getUndertowOptions().get(UndertowOptions.DECODE_URL, true);
+    }
+
+    private static void decodePath(HttpServerExchange exchange, String charset, StringBuilder sb) {
+        final boolean decodeSlash = exchange.getConnection().getUndertowOptions().get(
+                UndertowOptions.ALLOW_ENCODED_SLASH, false);
+        exchange.setRequestPath(URLUtils.decode(exchange.getRequestPath(), charset, decodeSlash, false, sb));
+        exchange.setRelativePath(URLUtils.decode(exchange.getRelativePath(), charset, decodeSlash, false, sb));
+        exchange.setResolvedPath(URLUtils.decode(exchange.getResolvedPath(), charset, decodeSlash, false, sb));
+    }
+
+    private static void decodeQueryString(HttpServerExchange exchange, String charset, StringBuilder sb) {
+        if (!exchange.getQueryString().isEmpty()) {
+            Map<String, Deque<String>> newParams = new TreeMap<>();
+            for (Map.Entry<String, Deque<String>> param : exchange.getQueryParameters().entrySet()) {
+                final Deque<String> newVales = new ArrayDeque<>(param.getValue().size());
+                for (String val : param.getValue()) {
+                    newVales.add(URLUtils.decode(val, charset, true, true, sb));
+                }
+                newParams.put(URLUtils.decode(param.getKey(), charset, true, true, sb), newVales);
+            }
+            exchange.getQueryParameters().clear();
+            exchange.getQueryParameters().putAll(newParams);
+        }
+    }
+
+    private static void decodePathTemplateMatch(HttpServerExchange exchange, String charset, StringBuilder sb) {
+        PathTemplateMatch pathTemplateMatch = exchange.getAttachment(PathTemplateMatch.ATTACHMENT_KEY);
+        if (pathTemplateMatch != null) {
+            Map<String, String> parameters = pathTemplateMatch.getParameters();
+            if (parameters != null) {
+                for (Map.Entry<String, String> entry : parameters.entrySet()) {
+                    entry.setValue(URLUtils.decode(entry.getValue(), charset, true, false, sb));
+                }
+            }
+        }
+    }
+
+    private static StringBuilder getStringBuilderForDecoding(HttpServerExchange exchange) {
+        if (exchange.isInIoThread()) {
```

Comment:
Requests are received on an IO thread, once they reach BlockingHandler (which we register in ConjureHandler) they will transition to a worker thread.

We're generally best off doing as much non-blocking initialization on the IO thread since they're meant to be constantly available and scale with CPU, which has a side-effect of reducing the stack depth on the task thread (where service methods are executed) for prettier stack traces.

Reply:
`We're generally best off doing as much non-blocking initialization on the IO thread since they're meant to be constantly available and scale with CPU` Is that always true ? Isn't there a point where ioThreads become the new bottleneck and reduce the possibility of doing some service reprioritization (which ofc we don't do, but just in theory) ?